src/threading/Manager.cc

```diff
@@ -1,33 +1,45 @@
+#include <sys/socket.h>
+#include <unistd.h>
 
 #include "Manager.h"
 #include "NetVar.h"
+#include "iosource/Manager.h"
 
 using namespace threading;
 
+void HeartbeatTimer::Dispatch(double t, int is_expire)
+	{
+	if ( is_expire )
+		return;
+
+	thread_mgr->SendHeartbeats();
+	thread_mgr->StartHeartbeatTimer();
+	}
+
 Manager::Manager()
 	{
 	DBG_LOG(DBG_THREADING, "Creating thread manager ...");
 
 	did_process = true;
 	next_beat = 0;
-	terminating = false;
-	SetIdle(true);
 	}
 
 Manager::~Manager()
 	{
 	if ( all_threads.size() )
 		Terminate();
+
+	if ( heartbeat_timer )
+		delete heartbeat_timer;
 	}
 
 void Manager::Terminate()
 	{
 	DBG_LOG(DBG_THREADING, "Terminating thread manager ...");
 
-	terminating = true;
-
 	// First process remaining thread output for the message threads.
```

Comment:
My guess is yes, because it flushes everything still in flight, i.e., data that would get lost otherwise.

Reply:
Removed the TODO